javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -12,31 +12,169 @@
 
 package java.util
 
+import java.lang.{reflect => jlr}
+
 import scala.scalajs.js
 
 import scala.annotation.tailrec
 
-import scala.reflect.ClassTag
-
 import ScalaOps._
 
 object Arrays {
 
+  /** Safe upcasting of an array, following run-time covariance. */
+  @inline private def arrayUpcast[A](a: Array[_ <: A]): Array[A] =
+    a.asInstanceOf[Array[A]]
+
+  @inline private def componentTypeOf[A](clazz: Class[_ <: Array[A]]): Class[_ <: A] =
+    clazz.getComponentType().asInstanceOf[Class[_ <: A]]
+
+  @inline private def componentTypeOf[A](a: Array[A]): Class[_ <: A] =
+    componentTypeOf(a.getClass())
+
+  /** A custom typeclass for the operations we need in `Arrays` to implement
+   *  the algorithms generically.
+   */
+  private sealed trait ArrayOps[A] {
+    def length(a: Array[A]): Int
+    def get(a: Array[A], i: Int): A
+    def set(a: Array[A], i: Int, v: A): Unit
+  }
+
+  private sealed trait ArrayCreateFromOps[A] extends ArrayOps[A] {
+    def createFrom(template: Array[A], length: Int): Array[A]
+  }
+
+  private sealed trait ArrayCreateOps[A] extends ArrayCreateFromOps[A] {
+    def create(length: Int): Array[A]
+
+    @inline def createFrom(template: Array[A], length: Int): Array[A] =
+      create(length)
```

Comment:
Rather than having `createFrom` like this, consider introducing a `TemplateArrayOps`: This will remove a lot of machinery and unclean separation of concerns from the `copyOf` implementation (notably that `copyOfImpl` takes a `ArrayCreateOps[_ <: T]` and not a `ArrayCreateOps[T]`).

The only usage of this is in the `sort` implementation. We can just pass the original array as a template.

This will make all the helpers above go away (or become local to `TemplateArrayOps`).

I realize that this will likely require to split out comparing in its own typeclass (or create `ComparatorTemplateArrayOps`). However, I do think this is cleaner anyways (can't we just use `java.util.Comparator`). Are there any optimization triggers that I have missed?

Reply:
I'm not sure what you have in mind, exactly. What would you put in `TemplateArrayOps`?

Reply:
> (notably that `copyOfImpl` takes a `ArrayCreateOps[_ <: T]` and not a `ArrayCreateOps[T]`)

That is a direct consequence of the signature of `copy`, which is
```scala
def copyOf[T <: AnyRef, U <: AnyRef](original: Array[U], newLength: Int,
    newType: Class[_ <: Array[T]]): Array[T] = {
```
It is given the ability to create some type of array `_ <: Array[T]`, which with run-time covariance means the same as an `Array[_ <: T]`, yet it returns an `Array[T]`.

Reply:
I don't think we could make the typeclass covariant in a way that is acceptable for the compile-time variance checking. The naive definition would be
```scala
trait ArrayCreateOps[+T] {
  def create(length: Int): Array[T]
}
```
but just that definition fails to compile with
```
-- Error: ----------------------------------------------------------------------
2 |  def create(length: Int): Array[T]
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |covariant type T occurs in invariant position in type (length: Int): Array[T] of method create
1 error found
```
We would have to make it return an `Array[_ <: T]` for the definition of the typeclass to compile, but then we would still be stuck in `copyOfImpl`.

Reply:
Right... But could we define `ClassArrayOps[A](clazz: Class[_ <: A])`?

I just feel like there are two different ways of handling array creation (`create` for `copyOf` and `createFrom` for `sort`) and it at least feels that we can use the same way for both of these.

Reply:
I've been trying to move things around and define things in different ways, but everything I tried seemed more complicated, when it worked at all, than the status quo.

> Right... But could we define `ClassArrayOps[A](clazz: Class[_ <: A])`?

I don't think we can do that in a way that preserves correctness. When given a `Class[Array[X]]` in `copyOf`, we must create an `Array[X]`. Not an `Array[_ <: X]`. If we immediately loosen the `clazz` to `Class[_ <: A]`, how can we preserve that guarantee?

> I just feel like there are two different ways of handling array creation (`create` for `copyOf` and `createFrom` for `sort`) and it at least feels that we can use the same way for both of these.

There *are* two different ways. `createFrom` is when you don't have a `Class[X]` of your array type, but you have a pre-existing template array that you can immitate. And `create` is when you don't have a template array, but you statically know the target type (primitives) or have a `Class[_]` of the expected array type.

Being able to `create` is a stronger, so it's in a subtrait of the one for `createFrom`. But requiring `create` doesn't work everywhere (like in `sort`).

I don't see how we could get rid of either. And I don't see why we should separate them.

Reply:
> There are two different ways.

Sorry, I wasn't clear. I meant there are two different ways to create an array from a template array.

`copyOf` (in some implementations), uses a `ClassArrayOps`. Example:
https://github.com/scala-js/scala-js/pull/4652/files#diff-6d2b3e0ec55f1ad6c27bad9b8065c42b3b55d2ec34a648feac4f9570cc361cdcR551

`sort` uses `createFrom`:
https://github.com/scala-js/scala-js/pull/4652/files#diff-6d2b3e0ec55f1ad6c27bad9b8065c42b3b55d2ec34a648feac4f9570cc361cdcR256

But when we trace back the calls, there is no reason the call site could not also create a `ClassArrayOps` just like the `copyOf` implementation.

That's what I meant by "there are two ways". Maybe there are three ways :) Of course we need the possibility to create a new array from a template and from a class tag.

> I don't think we can do that in a way that preserves correctness. When given a Class[Array[X]] in copyOf, we must create an Array[X]

My brain hurts :-/ I think the bottom line is, that I do not understand when the static typing reflects runtime typing rules and when it reflects static typing rules.

What follows is my writeup how I got to this conclusion:

Indeed, in `copyOf`, we get `Class[_ <: Array[T]]` (and not `Class[Array[_ <: T]]`).
https://github.com/scala-js/scala-js/pull/4652/files#diff-6d2b3e0ec55f1ad6c27bad9b8065c42b3b55d2ec34a648feac4f9570cc361cdcR555

Given `T <: AnyRef` (which in this case it is):
In terms of java typing (arrays are covariant), that's essentially the same (as long as there are no subclasses of arrays).
In terms of scala typing, it is not indeed.

Maybe I do not understand when we use which typing rules. For example why does `componentTypeOf` return `Class[_ <: A]` and not `Class[A]`?
https://github.com/scala-js/scala-js/pull/4652/files#diff-6d2b3e0ec55f1ad6c27bad9b8065c42b3b55d2ec34a648feac4f9570cc361cdcR29

Also, why does `copyOfImpl` take a `ArrayCreateOps[_ <: T]` and not a `tops: ArrayCreateOps[T]`? (which in turns requires the use of `arrayUpcast`)? 

**pause**, I realize I have asked this question before (in this thread). Answer:

> It is given the ability to create some type of array _ <: Array[T], which with run-time covariance means the same as an Array[_ <: T], yet it returns an Array[T].

--> Runtime typing rules apply.

Compare this to the original comment:

> I don't think we can do that in a way that preserves correctness. When given a Class[Array[X]] in copyOf, we must create an Array[X]. Not an Array[_ <: X].

--> Static typing rules apply. (otherwise `Array[_ <: X] <: Array[X]` given `X <: AnyRef`).